encodings/alp/src/alp/operator.rs

```diff
@@ -0,0 +1,117 @@
+        match self.patches() {
+            Some(_) => vortex_bail!("patched ALP kernel not implemented",),
+            None => {
+                if !self.all_valid() {
+                    vortex_bail!("ALP kernel does not yet handle nulls",);
+                }
+                match_each_alp_float_ptype!(self.ptype(), |A| {
+                    Ok(Box::new(UnpatchedALPKernel {
+                        encoded_vector_id,
+                        exponents: self.exponents(),
+                        _phantom: PhantomData::<A>::default(),
+                    }))
+                })
+            }
+        }
+    }
+}
+
+struct UnpatchedALPKernel<A> {
+    encoded_vector_id: VectorId,
+    exponents: Exponents,
+    _phantom: PhantomData<A>,
+}
+
+impl<A: ALPFloat> Kernel for UnpatchedALPKernel<A> {
+    fn step(
+        &mut self,
+        ctx: &KernelCtx,
+        selection: &BitView,
+        out: &mut VectorMut,
+    ) -> VortexResult<()> {
+        if selection.true_count() == 0 {
+            // Nothing to do, and no kernel state to update
+            return Ok(());
+        }
+
+        let encoded = ctx.input(self.encoded_vector_id);
+        let encoded_vec = encoded.as_primitive().downcast::<A::ALPInt>();
+        let encoded_buf = encoded_vec.elements().as_slice();
+
+        let decoded = out.as_primitive_mut().downcast::<A>();
+
+        match encoded.position() {
+            Position::InPlace => {
+                // TODO(ngates): tune the threshold
+                if selection.true_count() < (N / 8) {
+                    // Operate only over the selected elements, appending `true_count` elements
+                    unsafe {
+                        decoded
+                            .validity_mut()
+                            .append_n(true, selection.true_count())
+                    };
+                    unsafe { decoded.elements_mut().set_len(selection.true_count()) };
+                    let decoded_buf = unsafe { decoded.elements_mut() };
+
+                    let mut out_pos = 0;
+                    selection.iter_ones(|idx| {
+                        let encoded = unsafe { encoded_buf.get_unchecked(idx) };
+                        let decoded = A::decode_single(*encoded, self.exponents);
+                        *unsafe { decoded_buf.get_unchecked_mut(out_pos) } = decoded;
+                        out_pos += 1;
+                    })
+                } else {
+                    // Operate over all N elements, appending N elements
+                    assert_eq!(encoded_buf.len(), N);
+                    decoded.extend(
+                        encoded_buf
+                            .iter()
+                            .map(|e| A::decode_single(*e, self.exponents)),
+                    );
+                }
+            }
+            Position::Compact => {
+                // Loop over the compacted input elements
+                decoded.extend(
```

Comment:
Compact and "dense" InPlace both have the same code...

It hasn't really been documented, but not all input vectors have length == N. For those with len < N, we _know_ that the elements are compacted.

For those with len == N, we know that the elements are necessarily in-place.

So I wonder if we simply get rid of the Position enum entirely and we just look at input length + selection mask?